### DISCUSSION TITLE
How do I connect to PlanetScale using Go

### DISCUSSION THREAD

**Discussion by marwan-at-work** [planetscale/discussion#24] [Q&A] [answered]
Hi there 👋🏼 

I am wondering how I can connect my Go server to a PlanetScale database in a production environment such as Cloud Run. 

All of my digging so far shows that I need the "pscale" CLI as well as creating a localhost proxy and I'm hoping there's a better way where I can just connect with a Go library without having to set any of those dependencies up in my production environment. 

Thanks!

**Comment by Anders-PlanetScale** [comment]
Howdy Marwan,

For the time being you are correct that you will need to use  pscale CLI to access your databases. You can use either a localhost proxy created using the `pscale connect`command or use `pscale shell`  command to get a mysql command prompt.

Our engineers are looking into options to add a direct connection string. It is proving to be a common feature request, and we will follow up as we have more information.

**Comment by fatih** [comment, marked as answer]
Hi @marwan-at-work. Thanks for the feedback. Here is an example that you can use in your Go service. It uses our official `planetscale-go` Go package to fetch certs signed for your organization and then establish a secure connection to PlanetScale:

```go
package main

import (
	"context"
	"crypto/rand"
	"crypto/rsa"
	"crypto/tls"
	"crypto/x509"
	"database/sql"
	"errors"
	"flag"
	"fmt"
	"net"
	"os"
	"strconv"

	"github.com/go-sql-driver/mysql"
	ps "github.com/planetscale/planetscale-go/planetscale"
)

func main() {
	if err := realMain(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

// Run with command:
// go run main.go --token (cat ~/.config/planetscale/access-token) --org myorg --database mydb --branch mybranch
func realMain() error {
	token := flag.String("token", "", "PlanetScale Access Token")
	org := flag.String("org", "", "PlanetScale organization name")
	database := flag.String("database", "", "PlanetScale db name")
	branch := flag.String("branch", "", "PlanetScale branch name")

	flag.Parse()

	if *token == "" {
		return errors.New("--token is not set")
	}
	if *org == "" {
		return errors.New("--org is not set")
	}
	if *database == "" {
		return errors.New("--database is not set")
	}
	if *branch == "" {
		return errors.New("--branch is not set")
	}

	client, err := ps.NewClient(ps.WithAccessToken(*token))
	if err != nil {
		return err
	}

	pkey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return fmt.Errorf("couldn't generate private key: %s", err)
	}

	fmt.Println("==> fetching certs...")
	ctx := context.Background()
	cert, err := client.Certificates.Create(ctx, &ps.CreateCertificateRequest{
		Organization: *org,
		DatabaseName: *database,
		Branch:       *branch,
		PrivateKey:   pkey,
	})
	if err != nil {
		return err
	}
	rootCertPool := x509.NewCertPool()
	rootCertPool.AddCert(cert.CACert)

	serverName := fmt.Sprintf("%s.%s.%s.%s", *branch, *database, *org, cert.RemoteAddr)

	err = mysql.RegisterTLSConfig("planetscale", &tls.Config{
		RootCAs:      rootCertPool,
		Certificates: []tls.Certificate{cert.ClientCert},
		ServerName:   serverName,
		// we use InseruceSkipVerify with VerifyConnection, to make testing
		// easier on localhost and avoid the error:
		// 'invalid certificate name "localhost", expected "mybranch.mydb.myorg.region.psdb.cloud"'
		InsecureSkipVerify: true,
		VerifyConnection: func(cs tls.ConnectionState) error {
			opts := x509.VerifyOptions{
				Roots:         rootCertPool,
				Intermediates: x509.NewCertPool(),
			}
			for _, cert := range cs.PeerCertificates[1:] {
				opts.Intermediates.AddCert(cert)
			}
			_, err := cs.PeerCertificates[0].Verify(opts)
			return err
		},
	})
	if err != nil {
		return err
	}

	cfg := mysql.NewConfig()
	cfg.Addr = net.JoinHostPort(serverName, strconv.Itoa(cert.Ports.MySQL))
	cfg.Net = "tcp"
	cfg.TLSConfig = "planetscale"

	fmt.Println("==> connecting to the db ...")
	db, err := sql.Open("mysql", cfg.FormatDSN())
	if err != nil {
		return err
	}
	defer db.Close()

	// Open doesn't open a connection. Validate DSN data:
	err = db.Ping()
	if err != nil {
		return err
	}

	var version string
	err = db.QueryRow("SELECT VERSION()").Scan(&version)
	if err != nil {
		return err
	}

	fmt.Println("MySQL version:", version)
	return nil
}
```

Save this as a `main.go` file and then you can run it with:

```
$ go run demo.go --token (cat ~/.config/planetscale/access-token) --org planetscale --database planetscale --branch main
==> fetching certs...
==> connecting to the db ...
MySQL version: 8.0.23
```

We're planning to add a helper function to the planetscale-go Go package to simplify this process. Let me know if you have other questions.

**Reply by marwan-at-work** [reply]
@fatih this is awesome and works well 🙏🏼 I'm super impressed by how fast roundtrips are from my local laptop! 

My last question is: it seems that the token must be the one that was browser-generated aka `cat ~/.config/planetscale/access-token` -- Is there a way to use a more proper token such as the one from `pscale service-token create` ? 

Thanks again

**Reply by ghazninattarshah** [reply]
> Hi @marwan-at-work. Thanks for the feedback. Here is an example that you can use in your Go service. It uses our official `planetscale-go` Go package to fetch certs signed for your organization and then establish a secure connection to PlanetScale:
> 
> ```go
> package main
> 
> import (
> 	"context"
> 	"crypto/rand"
> 	"crypto/rsa"
> 	"crypto/tls"
> 	"crypto/x509"
> 	"database/sql"
> 	"errors"
> 	"flag"
> 	"fmt"
> 	"net"
> 	"os"
> 	"strconv"
> 
> 	"github.com/go-sql-driver/mysql"
> 	ps "github.com/planetscale/planetscale-go/planetscale"
> )
> 
> func main() {
> 	if err := realMain(); err != nil {
> 		fmt.Fprintln(os.Stderr, err)
> 		os.Exit(1)
> 	}
> }
> 
> // Run with command:
> // go run main.go --token (cat ~/.config/planetscale/access-token) --org myorg --database mydb --branch mybranch
> func realMain() error {
> 	token := flag.String("token", "", "PlanetScale Access Token")
> 	org := flag.String("org", "", "PlanetScale organization name")
> 	database := flag.String("database", "", "PlanetScale db name")
> 	branch := flag.String("branch", "", "PlanetScale branch name")
> 
> 	flag.Parse()
> 
> 	if *token == "" {
> 		return errors.New("--token is not set")
> 	}
> 	if *org == "" {
> 		return errors.New("--org is not set")
> 	}
> 	if *database == "" {
> 		return errors.New("--database is not set")
> 	}
> 	if *branch == "" {
> 		return errors.New("--branch is not set")
> 	}
> 
> 	client, err := ps.NewClient(ps.WithAccessToken(*token))
> 	if err != nil {
> 		return err
> 	}
> 
> 	pkey, err := rsa.GenerateKey(rand.Reader, 2048)
> 	if err != nil {
> 		return fmt.Errorf("couldn't generate private key: %s", err)
> 	}
> 
> 	fmt.Println("==> fetching certs...")
> 	ctx := context.Background()
> 	cert, err := client.Certificates.Create(ctx, &ps.CreateCertificateRequest{
> 		Organization: *org,
> 		DatabaseName: *database,
> 		Branch:       *branch,
> 		PrivateKey:   pkey,
> 	})
> 	if err != nil {
> 		return err
> 	}
> 	rootCertPool := x509.NewCertPool()
> 	rootCertPool.AddCert(cert.CACert)
> 
> 	serverName := fmt.Sprintf("%s.%s.%s.%s", *branch, *database, *org, cert.RemoteAddr)
> 
> 	err = mysql.RegisterTLSConfig("planetscale", &tls.Config{
> 		RootCAs:      rootCertPool,
> 		Certificates: []tls.Certificate{cert.ClientCert},
> 		ServerName:   serverName,
> 		// we use InseruceSkipVerify with VerifyConnection, to make testing
> 		// easier on localhost and avoid the error:
> 		// 'invalid certificate name "localhost", expected "mybranch.mydb.myorg.region.psdb.cloud"'
> 		InsecureSkipVerify: true,
> 		VerifyConnection: func(cs tls.ConnectionState) error {
> 			opts := x509.VerifyOptions{
> 				Roots:         rootCertPool,
> 				Intermediates: x509.NewCertPool(),
> 			}
> 			for _, cert := range cs.PeerCertificates[1:] {
> 				opts.Intermediates.AddCert(cert)
> 			}
> 			_, err := cs.PeerCertificates[0].Verify(opts)
> 			return err
> 		},
> 	})
> 	if err != nil {
> 		return err
> 	}
> 
> 	cfg := mysql.NewConfig()
> 	cfg.Addr = net.JoinHostPort(serverName, strconv.Itoa(cert.Ports.MySQL))
> 	cfg.Net = "tcp"
> 	cfg.TLSConfig = "planetscale"
> 
> 	fmt.Println("==> connecting to the db ...")
> 	db, err := sql.Open("mysql", cfg.FormatDSN())
> 	if err != nil {
> 		return err
> 	}
> 	defer db.Close()
> 
> 	// Open doesn't open a connection. Validate DSN data:
> 	err = db.Ping()
> 	if err != nil {
> 		return err
> 	}
> 
> 	var version string
> 	err = db.QueryRow("SELECT VERSION()").Scan(&version)
> 	if err != nil {
> 		return err
> 	}
> 
> 	fmt.Println("MySQL version:", version)
> 	return nil
> }
> ```
> 
> Save this as a `main.go` file and then you can run it with:
> 
> ```
> $ go run demo.go --token (cat ~/.config/planetscale/access-token) --org planetscale --database planetscale --branch main
> ==> fetching certs...
> ==> connecting to the db ...
> MySQL version: 8.0.23
> ```
> 
> We're planning to add a helper function to the planetscale-go Go package to simplify this process. Let me know if you have other questions.

That was exactly what I was looking for. thanks @fatih for the sample

**Comment by fatih** [comment]
To make it easy to connect to Go applications, I added a helper method to our Go API package. Here is the PR: https://github.com/planetscale/planetscale-go/pull/61

Once this PR is merged, it'll be as simple as importing the `planetscale/dbutil` and use it like:

```go
package main

import (
	"context"
	"crypto/rand"
	"crypto/rsa"
	"crypto/tls"
	"crypto/x509"
	"database/sql"
	"errors"
	"flag"
	"fmt"
	"net"
	"os"
	"os/signal"
	"strconv"

	"github.com/go-sql-driver/mysql"
	"github.com/planetscale/planetscale-go/planetscale"
	"github.com/planetscale/planetscale-go/planetscale/dbutil"
)

func main() {
	if err := realMain(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

// Run with command:
// go run main.go --token (cat ~/.config/planetscale/access-token) --org myorg --database mydb --branch mybranch
func realMain() error {
	token := flag.String("token", "", "PlanetScale Access Token")
	org := flag.String("org", "", "PlanetScale organization name")
	database := flag.String("database", "", "PlanetScale db name")
	branch := flag.String("branch", "", "PlanetScale branch name")

	flag.Parse()

	if *token == "" {
		return errors.New("--token is not set")
	}
	if *org == "" {
		return errors.New("--org is not set")
	}
	if *database == "" {
		return errors.New("--database is not set")
	}
	if *branch == "" {
		return errors.New("--branch is not set")
	}

	client, err := planetscale.NewClient(planetscale.WithAccessToken(*token))
	if err != nil {
		return err
	}

	db, err := dbutil.Dial(ctx, &dbutil.DialConfig{
		Organization: *org,
		Database:     *database,
		Branch:       *branch,
		Client:       client,
	})
	if err != nil {
		return err
	}

	var version string
	err = db.QueryRow("SELECT VERSION()").Scan(&version)
	if err != nil {
		return err
	}

	fmt.Println("MySQL version:", version)
	return nil
}
```

**Reply by fatih** [reply]
This change is now merged and available with the `planetscale-go` version https://github.com/planetscale/planetscale-go/releases/tag/v0.27.0

Please let me know if you see any issues.

**Reply by croaky** [reply]
Using the `dbutil` version from the planetscale-go README, I experience the same as @marwan-at-work's earlier report: I'm able to connect with `PLANETSCALE_TOKEN=$(cat ~/.config/planetscale/access-token)` but not with `PLANETSCALE_TOKEN` equal to a service token per [these docs](https://docs.planetscale.com/tutorial/connect-any-application#authentication-in-production):

```
pscale service-token create
pscale service-token add-access your-token-name connect_production_branch --database your-db-name
```

When I look for additional access controls I could attenuate the token with, it looks like the options are `TODO`:

```
% pscale service-token add-access --help                                       
add access to a service token in the organization

Usage:
  pscale service-token add-access <token> <access> <access> ... [flags]

Examples:
The add-access command grants a service token specific access on a specific database.

For example, to give a service token the ability to create, read and delete branches on a specific database:

  pscale token add-access <token id> read_branch delete_branch create_branch --database <database name>

For a complete list of the access permissions that can be granted to a token, see: TODO.
```

CLI version:

```
pscale version 0.39.0 (build date: 2021-05-24T12:03:51Z commit: b321aeb)
```

**Reply by fatih** [reply]
What problem are you seeing? Could you open a bug report to the repo (planetscale-go) with the code you’re using. That would be eaiser for us to debug.

**Reply by croaky** [reply]
Opened at https://github.com/planetscale/planetscale-go/issues/65
